Keda.Cosmosdb.Scaler/src/Services/CosmosDBExternalScaler.cs

```diff
@@ -0,0 +1,173 @@
+            {
+                MetricName = NormalizeString(string.Format("{0}-{1}-{2}-{3}", "azure-cosmosDB",
+                                        request.ScalerMetadata[CosmosDBTriggerMetadata.AccountName],
+                                        request.ScalerMetadata[CosmosDBTriggerMetadata.DatabaseName],
+                                        request.ScalerMetadata[CosmosDBTriggerMetadata.CollectionName])),
+                TargetSize = 1
+            });
+            return Task.FromResult(resp);
+        }
+
+        public override async Task<GetMetricsResponse> GetMetrics(GetMetricsRequest request, ServerCallContext context)
+        {
+            var resp = new GetMetricsResponse();
+            try
+            {
+                var trigger = CreateTriggerFromMetadata(request.ScaledObjectRef.ScalerMetadata);
+                long workToBeDone = await GetEstimatedWork(trigger);
+
+                var metricName = request.ScaledObjectRef.ScalerMetadata["MetricName"];
+                resp.MetricValues.Add(new MetricValue
+                {
+                    MetricName = metricName,
+                    MetricValue_ = workToBeDone
+                });
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Error generating GetMetricsResponse");
+                throw ex;
+            }
+            return resp;
+        }
+
+        public CosmosDBTrigger CreateTriggerFromMetadata(MapField<string, string> scalerMetadata)
+        {
+            var lease = new CosmosDBLease()
+            {
+                LeasesCosmosDBConnectionString = scalerMetadata[CosmosDBLeaseMetadata.LeasesCosmosDBConnectionString],
+                LeaseDatabaseName = scalerMetadata[CosmosDBLeaseMetadata.LeaseDatabaseName],
+                LeaseCollectionName = scalerMetadata[CosmosDBLeaseMetadata.LeaseCollectionName]
+            };
+
+            var trigger = new CosmosDBTrigger()
+            {
+                CollectionName = scalerMetadata[CosmosDBTriggerMetadata.CollectionName],
+                DatabaseName = scalerMetadata[CosmosDBTriggerMetadata.DatabaseName],
+                CosmosDBConnectionString = scalerMetadata[CosmosDBTriggerMetadata.CosmosDBConnectionString],
+                AccountName = string.Empty,
+                Lease = lease
+            };
+
+            if (scalerMetadata.TryGetValue(CosmosDBTriggerMetadata.AccountName, out string accountName))
+            {
+                trigger.AccountName = accountName;
+            }
+
+            return trigger;
+        }
+
+        private async Task<long> GetEstimatedWork (CosmosDBTrigger trigger)
+        {
+            try
+            {
+                var builder = CreateChangeFeedProcessorBuilder(trigger);
+                var workEstimator = await builder.BuildEstimatorAsync();
+                long estimatedRemainingWork = await workEstimator.GetEstimatedRemainingWork();

```

Comment:
We're creating a new client every time, shouldn't we:
- Inject the client instead of building it every time here?
- Re-use clients to optimize performance?

Reply:
What I would do is:
- Create a `CosmosDbRepository` which has a `GetEstimatedWork` method
- `CosmosDbRepository` uses the builder that is injected to connect and determine the remaining work
- `CosmosDbExternalScaler` uses injected `CosmosDbRepository` to determine the work

Reply:
A single instance of this pod can serve requests for different function apps, so it needs to connect to different cosmos DB accounts based on the incoming request. Due to this, I cannot inject the builder at startup since we only get the cosmos DB account details at request time. 
However, I agree that we shouldn't be recreating the builder every time for the same cosmos DB account. Will an in- memory cache of builders for different cosmos DB accounts be acceptable to you?

Reply:
Sure, but I'd still move out the logic into repos so that the scaler just calls that and orchestrates the metric retrieval. Make sense to you?

Reply:
Thanks that's already done in the latest iteration. Please take a look.

Reply:
The inmemory cache has been added. @tomkerkhove Please take a look when you get a chance.